### PR TITLE
Redesign part 2

### PR DESCRIPTION
**Pull request by PrzeChoj**
Builder Design Pattern on `gips` class.

### CODE REVIEW

File: tests/testthat/test_find_gips.R [hunk 1]
```diff
@@ -12,7 +12,7 @@ test_that("Process proper parameters", {
 test_that("Handle improper parameters", {
   expect_error(best_growth(S=matrix_invariant_by_example_perm, number_of_observations=number_of_observations, max_iter=Inf,
                            start_perm=example_perm, show_progress_bar=TRUE))
-  expect_error(gips(matrix_invariant_by_example_perm, number_of_observations, 10, return_probabilities = TRUE, optimizer = "BG"))
+  expect_error(find_gips(gips(matrix_invariant_by_example_perm/5, number_of_observations), 10, return_probabilities = TRUE, optimizer = "BG"))  # TODO(Do not divide this by 5)
```
Comment (MrDomani) [inline tests/testthat/test_find_gips.R]: ?

File: tests/testthat/test_gips_class.R [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+g2 <- gips(S, number_of_observations, perm = custom_perm2)
+
+test_that('Setting custom permutation in gips constructor works',{
+  expect_identical(custom_perm1, g1[[1]])
+  
+  expect_identical(gips_perm(custom_perm2, ncol(S)), g2[[1]])
+})
+
+
+test_that('Properly validate the gips class',{
```
Comment (MrDomani) [inline tests/testthat/test_gips_class.R:30]: Wow, that's a lot of tests

File: R/gips_class.R [hunk 3]
```diff
@@ -0,0 +1,573 @@
+#' Constructor of the `gips` class.
+#' 
+#' Create the `gips` object.
+#' This object will consists data and all other information needed to find the invariant group.
+#' The optimization itself will not be performed. To do it, one have to call the TODO() function. See examples below.
```
Comment (MrDomani) [inline R/gips_class.R]: is TODO() function ready?

File: R/gips_class.R [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+#'
+#' Only intended for low-level use.
+#'
+#' @param list_of_gips_perm list with a single element of class `gips_perm`. The base object for the `gips` class.
+#' @param optimization_info NULL or the list with information about the optimization process.
+#'
+#' @export
+new_gips <- function(list_of_gips_perm, S, number_of_observations, delta,
+                     D_matrix, optimization_info){
+  stopifnot(is.list(list_of_gips_perm),
```
Comment (MrDomani) [inline R/gips_class.R]: Why not `rlang::abort`?

File: R/gips_class.R [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+
+# TODO(Documentation; I think @describeIn is intended for methods...)
+#' @describeIn gips Validator
+#'
+#' Only intended for low-level use.
+#'
+#' @param g Element to be checked if it is of class `gips`.
+#'
+#' @export
+validate_gips <- function(g){
```
Comment (MrDomani) [inline R/gips_class.R:105]: Not gonna lie, I did not look into code of this function

File: R/find_gips.R [hunk 6]
```diff
@@ -1,18 +1,14 @@
-#' Metropolis-Hastings algorithm
```
Comment (MrDomani) [inline R/find_gips.R]: Not gonna lie, I just glanced over this file

File: tests/testthat/test_find_gips.R (hunk 1 above)
Comment (PrzeChoj) [reply tests/testthat/test_find_gips.R]: I guess You are confused by the "TODO". There was a problem with this matrix that I did not understand until You explained this to me today. I have solved this problem and deleted the TODO in the commit 0b409b81109395b4113446d10c1fc76016083d59.

File: R/gips_class.R (hunk 3 above)
Comment (PrzeChoj) [reply R/gips_class.R]: Yea, forgot to reference. It is up in 3ba879e48dbfdee98439caa87ddab62b4c0ef9b1.

File: R/gips_class.R (hunk 4 above)
Comment (PrzeChoj) [reply R/gips_class.R]: Yea, forgot to change. It is up in 3ba879e48dbfdee98439caa87ddab62b4c0ef9b1.

**Pull request merged**